pkg/gofr/datasource/surrealdb/surrealdb.go

```diff
@@ -0,0 +1,543 @@
+	if l, ok := customlogger.(Logger); ok {
+		c.logger = l
+	}
+}
+
+// UseMetrics sets a custom metrics recorder for the Client.
+func (c *Client) UseMetrics(metrics interface{}) {
+	if m, ok := metrics.(Metrics); ok {
+		c.metrics = m
+	}
+}
+
+// UseTracer sets a custom tracer for the Client.
+func (c *Client) UseTracer(tracer interface{}) {
+	if t, ok := tracer.(trace.Tracer); ok {
+		c.tracer = t
+	}
+}
+
+// Connect establishes a connection to the SurrealDB server using the client's configuration.
+var (
+	errNoDatabaseInstance       = errors.New("failed to connect to SurrealDB: no valid database instance")
+	errInvalidCredentialsConfig = errors.New("both username and password must be provided")
+	errEmbeddedDBNotEnabled     = errors.New("embedded database not enabled")
+	errInvalidConnectionURL     = errors.New("invalid connection URL")
+)
+
+// NewDB creates a new SurrealDB client.
+func NewDB(connectionURL string) (con connection.Connection, err error) {
+	u, err := url.ParseRequestURI(connectionURL)
+	if err != nil {
+		return nil, err
+	}
+
+	scheme := u.Scheme
+
+	newParams := connection.NewConnectionParams{
+		Marshaler:   models.CborMarshaler{},
+		Unmarshaler: models.CborUnmarshaler{},
+		BaseURL:     fmt.Sprintf("%s://%s", u.Scheme, u.Host),
+		Logger:      logger.New(slog.NewTextHandler(os.Stdout, nil)),
+	}
+
+	if scheme == "http" || scheme == https {
+		con = connection.NewHTTPConnection(newParams)
+	} else if scheme == ws || scheme == "wss" {
+		con = connection.NewWebSocketConnection(newParams)
+	} else if scheme == "memory" || scheme == "mem" || scheme == "surrealkv" {
+		return nil, fmt.Errorf("%w", errEmbeddedDBNotEnabled)
+	} else {
+		return nil, fmt.Errorf("%w", errInvalidConnectionURL)
+	}
+
+	err = con.Connect()
+	if err != nil {
+		return nil, err
+	}
+
+	return con, nil
+}
+
+// Connect establishes a connection to the SurrealDB database.
+func (c *Client) Connect() {
+	endpoint := c.buildEndpoint()
+	err := c.connectToDatabase(endpoint)
+
+	if err != nil {
+		return
+	}
+
+	err = c.setupNamespaceAndDatabase()
+	if err != nil {
+		return
+	}
+
+	err = c.authenticate()
+	if err != nil {
+		return
+	}
+
+	c.logger.Debugf("successfully connected to SurrealDB")
+}
+
+// buildEndpoint constructs the SurrealDB endpoint based on the configuration.
+func (c *Client) buildEndpoint() string {
+	scheme := ws
+	if c.config.TLSEnabled {
+		scheme = https
+	}
+
+	return fmt.Sprintf("%s://%s:%d", scheme, c.config.Host, c.config.Port)
+}
+
+// connectToDatabase handles the connection to SurrealDB and returns an error if failed.
+func (c *Client) connectToDatabase(endpoint string) error {
+	c.logger.Debugf("connecting to SurrealDB at %s", endpoint)
+
+	db, err := NewDB(endpoint)
+	if err != nil {
+		c.logError("failed to connect to SurrealDB", err)
+		return err
+	}
+
+	if db == nil {
+		c.logError("failed to connect to SurrealDB: no valid database instance", nil)
+		return errNoDatabaseInstance
+	}
+
+	c.db = db
+
+	return nil
```

Comment:
I would have done this

```suggestion
	var err error
	c.db, err = NewDB(endpoint)
	if err != nil {
		c.logError("failed to connect to SurrealDB", err)
		return err
	}

	if c.db == nil {
		c.logError("failed to connect to SurrealDB: no valid database instance", nil)
		return errNoDatabaseInstance
	}

	return nil
```